shell/platform/darwin/macos/framework/Source/FlutterApplication.mm

```diff
@@ -0,0 +1,86 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import "flutter/shell/platform/darwin/macos/framework/Headers/FlutterApplication.h"
+#include "embedder.h"
+#include "shell/platform/darwin/macos/framework/Headers/FlutterAppDelegate.h"
+#include "shell/platform/darwin/macos/framework/Source/FlutterAppDelegate_internal.h"
+
+#ifndef FLUTTER_RELEASE
+namespace {
+static bool warned_about_flutter_application = false;
+}  // namespace
+#endif  // !FLUTTER_RELEASE
+
+// An NSApplication subclass that implements overrides necessary for some
+// Flutter features, like application lifecycle handling.
+@interface FlutterApplication () {
+}
+@end
+
+@implementation FlutterApplication
+
+// Initialize NSApplication using the custom subclass.  Check whether NSApp was
+// already initialized using another class, because that would break some
+// things. Warn about the mismatch only once, and only in debug builds.
++ (NSApplication*)sharedApplication {
+  NSApplication* app = [super sharedApplication];
+
+  // +sharedApplication initializes the global NSApp, so if we're delivering
+  // something other than a FlutterApplication, warn the developer once.
+#ifndef FLUTTER_RELEASE
+  if (!warned_about_flutter_application && ![NSApp isKindOfClass:self]) {
```

Comment:
I wonder if rather than testing for the use of the FlutterApplication class, we can somehow test they've overridden that particular method. NSApplication itself will include a default implementation though. :/